pkg/apis/ceph.rook.io/v1/topic.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"net/url"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// compile-time assertions ensures CephBucketTopic implements webhook.Validator so a webhook builder
+// will be registered for the validating webhook.
+var _ webhook.Validator = &CephBucketTopic{}
+
+func ValidateHTTPSpec(s *HTTPEndpointSpec) error {
+	return nil
+}
+
+func ValidateAMQPSpec(s *AMQPEndpointSpec) error {
+	if s.Exchange == "" {
+		return errors.New("missing AMQP exchange")
+	}
+	return nil
+}
+
+func ValidateKafkaSpec(s *KafkaEndpointSpec) error {
+	return nil
+}
+
+// ValidateTopicSpec validate the bucket notification topic arguments
+func ValidateTopicSpec(t *CephBucketTopic) error {
+	if t.Name == "" {
+		return errors.New("missing name")
+	}
+	if t.Spec.Endpoint == "" {
+		return errors.New("missing endpoint")
+	}
+	var urlSchema string
+	endpointURL, err := url.Parse(t.Spec.Endpoint)
+	if err != nil {
+		return err
+	}
+	urlSchema = strings.ToLower(endpointURL.Scheme)
+	validSchemas := []string{"amqp", "amqps", "http", "kafka"}
+	if sort.SearchStrings(validSchemas, urlSchema) == len(validSchemas) {
```

Comment:
my use of "SearchStrings" here is wrong.
anyway, changed the way validation is done, so we allow for empty endpoint specs in case that we have defaults for all parameters. so I'm not using this functions anymore